src/garage/_primitives.py

```diff
@@ -0,0 +1,152 @@
+"""Policy interface."""
+import abc
+
+
+class Primitive(abc.ABC):
+    """Primitive interface.
+
+    Args:
+        name (str): Name for logs, variable scopes, etc.
+        spec (garage.InOutSpec): Input/output specification this primitive will
+            satisfy.
+
+    """
+
+    def __init__(self, name, spec):
+        self._name = name
+        self._spec = spec
+
+    @abc.abstractmethod
+    def get_parameters(self):
+        """Get parameters of this primitive.
+
+        Returns:
+            dict[str, np.ndarray]: A dictionary mapping from names to values of
+                each parameter.
+
+        """
+
+    @abc.abstractmethod
+    def set_parameters(self, parameters):
+        """Set the parameters of this primtive.
+
+        Args:
+            parameters (dict[str, np.ndarray]): A dictionary mapping from names
+                to values of each parameter.
+
+        """
+
+    @property
+    def spec(self):
+        """InOutSpec: The input/output spec for this primitive."""
+        return self._spec
+
+    @property
+    def name(self):
+        """str: Name of this primitive."""
+        return self._name
+
+    @abc.abstractmethod
+    def clone(self, name):
+        """Create a fully-independent copy of this primitive.
+
+        Args:
+            name (str): Name for the cloned instance.
+
+        Returns:
+            Primitive: A fully-independent copy of this primitive.
+
+        """
+
+    @abc.abstractmethod
+    def __getstate__(self):
+        """See Object.__getstate__.
+
+        Returns:
+            dict: Pickling state for the primitive
+
+        """
+
+    @abc.abstractmethod
+    def __setstate__(self, state):
+        """See Object.__setstate__.
+
+        Args:
+            state (dict): Unpickled state of this primitive.
+
+        """
+
+
+class Policy(Primitive):
+    """Policy interface.
+
+    Args:
+        name (str): Name for logs, variable scopes, etc.
+        env_spec (garage.envs.EnvSpec): Environment spec this Policy
+            will satisfy.
+
+    """
+
+    def __init__(self, name, env_spec):
+        super().__init__(name, env_spec.transpose())
```

Comment:
what is a spec transpose?

Reply:
it is defined above

spec(in_space, out_space).transpose() == spec(out_space, in_space)